editor/plugins/tool_button_editor_plugin.cpp

```diff
@@ -0,0 +1,77 @@
+/* permit persons to whom the Software is furnished to do so, subject to  */
+/* the following conditions:                                              */
+/*                                                                        */
+/* The above copyright notice and this permission notice shall be         */
+/* included in all copies or substantial portions of the Software.        */
+/*                                                                        */
+/* THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,        */
+/* EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF     */
+/* MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. */
+/* IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY   */
+/* CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,   */
+/* TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE      */
+/* SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.                 */
+/**************************************************************************/
+
+#include "tool_button_editor_plugin.h"
+#include "editor/editor_node.h"
+#include "editor/editor_property_name_processor.h"
+
+bool ToolButtonInspectorPlugin::can_handle(Object *p_object) {
+	Ref<Script> scr = Object::cast_to<Script>(p_object->get_script());
+	return scr.is_valid() && (*scr)->is_tool();
+}
+
+void ToolButtonInspectorPlugin::update_action_icon(Button *action_button) {
+	action_button->set_icon(action_button->get_theme_icon(action_icon, SNAME("EditorIcons")));
+}
+
+void ToolButtonInspectorPlugin::commit_action(Object *p_object, StringName p_method_name) {
+	if (!p_object->has_method(p_method_name)) {
+		print_error(vformat("Tool button method is invalid. Could not find method '%s' on %s", p_method_name, p_object->get_class_name()));
+		return;
+	}
+	p_object->call(p_method_name, EditorUndoRedoManager::get_singleton());
+}
+
+bool ToolButtonInspectorPlugin::parse_property(Object *p_object, const Variant::Type p_type, const String &p_path, const PropertyHint p_hint, const String &p_hint_text, const BitField<PropertyUsageFlags> p_usage, const bool p_wide) {
+	if (p_type == Variant::CALLABLE) {
+		if (p_usage.has_flag(PROPERTY_USAGE_INTERNAL) && p_usage.has_flag(PROPERTY_USAGE_EDITOR)) {
+			Button *action_button = EditorInspector::create_inspector_action_button(EditorPropertyNameProcessor::get_singleton()->process_name(p_path, EditorPropertyNameProcessor::STYLE_CAPITALIZED));
+
+			PackedStringArray split = p_hint_text.split(",");
+			if (split.size() > 2) {
+				String icon = split[2];
+				action_icon = StringName(icon);
+				action_button->connect(SNAME("theme_changed"), callable_mp(this, &ToolButtonInspectorPlugin::update_action_icon).bind(action_button));
+				action_button->connect(SNAME("tree_entered"), callable_mp(this, &ToolButtonInspectorPlugin::update_action_icon).bind(action_button));
```

Comment:
`tree_entered` is not necessary, as `theme_changed` is now automatically received on tree enter.